platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r812/CustomTestEventsCrossVersionSpec.groovy

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.integtests.tooling.r812
+
+
+import org.gradle.integtests.tooling.CustomTestEventsFixture
+import org.gradle.integtests.tooling.fixture.ProgressEvents
+import org.gradle.integtests.tooling.fixture.TargetGradleVersion
+import org.gradle.integtests.tooling.fixture.ToolingApiSpecification
+import org.gradle.integtests.tooling.fixture.ToolingApiVersion
+import org.gradle.tooling.ProjectConnection
+import org.gradle.tooling.events.OperationType
+
+@ToolingApiVersion(">=8.8")
+@TargetGradleVersion(">=8.12")
+class CustomTestEventsCrossVersionSpec extends ToolingApiSpecification implements CustomTestEventsFixture {
+    ProgressEvents events = ProgressEvents.create()
+
+    @Override
+    ProgressEvents getEvents() {
+        return events
+    }
+
+    def "reports custom test events (flat)"() {
+        given:
+        buildFile("""
+            import java.time.Instant
+
+            abstract class CustomTestTask extends DefaultTask {
+                @Inject
+                abstract TestEventService getTestEventService()
+
+                @TaskAction
+                void runTests() {
+                    try (def generator = getTestEventService().generateTestEvents("Custom test root")) {
+                        generator.started(Instant.now())
+                        try (def myTest = generator.createAtomicNode("MyTestInternal", "My test!")) {
```

Comment:
`createTest`/`createAtomic` and `createComposite` read kind of weirdly to me? I tried `createForTest` and `createForComposite` too, but in the end I think they should be thinking about the tree nature of the tests. It will be seen in reports and IDEs.

Reply:
I still think `Node` is mixing levels of abstraction here in a weird way.  I agree with you that just `createTest`/`createAtomic` isn't that great either.

While they should understand that the output format is tree-based, I don't know if I agree that callers should be thinking too much about the tree structure on the input side of things.  They are just reporting data points as they occur, and structuring them seems like something the API just handles internally.

I think something more like:

```
reporter.reportTestEvent() // atomic
reporter.reportTestGroup() // composite
```

is closer to what I would expect to see here.